### ISSUE TITLE
Gradients of Sinkhorn divergence without having to solve linear system

### ISSUE THREAD

**Issue by jaschau** [ott-jax/ott#415]
First of all, I want to thank you for this great library!

I have a use-case where I want to use the Sinkhorn divergence as an optimization criterion which is optimized with respect to the points x (with y fixed) of a PointCloud and with respect to the marginals. 
My understanding is that the library currently implements the most generic method for implicit differentiation outlined, e.g., in [1] which allows differentiating the optimal transport plan or the potentials. This has the downside that it requires the solution of a linear system. For my use-case, however, where I do not need the derivative of the optimal transport plan but only the derivative of the Sinkhorn divergence, I believe that the method detailed in [2] which works with the dual form of the Sinkhorn objective is computationally more efficient as it avoids solving a linear system.

Is my understanding outlined above correct? If so, I would be interested in implementing [2] as an alternative to the current approach.  Could this be done, e.g., by providing an alternative `ImplicitDiff` implementation? Or should I just create a function `my_sinkhorn_divergence(x, y, a, b)` with a custom gradient rule that implements Eq. 26, 27 from [2]?

Thanks a lot for your help!

[1] A Unified Framework for Implicit Sinkhorn Differentiation, https://arxiv.org/pdf/2205.06688.pdf
[2] Interpolating between Optimal Transport and MMD using Sinkhorn Divergences, https://arxiv.org/pdf/1810.08278.pdf

**Comment by jaschau** [comment]
I now discovered the `danskin` parameter in the Sinkhorn class. Does setting `danskin=True` actually already achieve what I asked for above?

**Comment by michalk8** [comment]
Hi @jaschau , indeed, specifying `use_danskin=True` will get you the desired results. In case of Sinkhorn divergence, you can pass it as:
```python
from ott.tools import sinkhorn_divergence
from ott.geometry import pointcloud

def f(x, y, a, b):
    out = sinkhorn_divergence.sinkhorn_divergence(
        pointcloud.PointCloud, x, y, a=a, b=b,
        sinkhorn_kwargs={"use_danskin": True}
    )
    return out.divergence, out

grad_a, out = jax.jit(jax.grad(f, has_aux=True, argnums=2))(x[:, None], y[:, None], a, b)
# f_i - p_i; eq. 26
assert jnp.isclose(grad_a, (out.potentials[0][0] - out.potentials[1][0])).all()
```

**Comment by marcocuturi** [comment]
Hi @jaschau 

Thanks for your interest in OTT and very excited to hear you are enjoying using it! Thanks a lot for raising this discussion.

as you noticed, `use_danskin` is the important flag here.

You should look in particular into the following [set_cost](https://ott-jax.readthedocs.io/en/latest/_modules/ott/solvers/linear/sinkhorn.html#SinkhornOutput.set_cost), which places a `stop_gradient` at the right spot.

Michal is right to mention that you can pass `use_danskin` as a `sinkhorn_kwargs`, but you don't really need to, it's already resolved as `True` by default, and this shows when computing `reg_ot_cost`. The same 3 `reg_ot_cost` variables corresponding to 3 different problems are used by [sinkhorn_divergence](https://ott-jax.readthedocs.io/en/latest/_autosummary/ott.tools.sinkhorn_divergence.sinkhorn_divergence.html#ott.tools.sinkhorn_divergence.sinkhorn_divergence), in the following block:

```
div = (
      out_xy.reg_ot_cost - 0.5 * (out_xx.reg_ot_cost + out_yy.reg_ot_cost) +
      0.5 * geometry_xy.epsilon * (jnp.sum(a) - jnp.sum(b)) ** 2
  )
```

Using this Danskin approx has been around, to the best of my knowledge, since https://proceedings.mlr.press/v32/cuturi14.html

To summarize what ott does: if you compute the gradients of `reg_ot_cost` (or stuff derived from it, e.g. `ent_reg_cost`) then you'll use Danskin automatically. Note that Danskin is **wrong** for higher order derivatives (e.g. Hessian), which is illustrated in [this nb](https://ott-jax.readthedocs.io/en/latest/tutorials/notebooks/Hessians.html) where we need to pass `use_danskin=False`.

If you compute the gradients of anything else (e.g. `primal_cost`, as in Luise's paper, see https://github.com/ott-jax/ott/blob/137fd3a656b6a9e647d33ba8f7183ee05a42848f/src/ott/solvers/linear/sinkhorn_lr.py#L259) then your computational graph will include the `f` and `g` optimal solutions, that will be then by default differentiated implicitly.

W.r.t implicit differentiation the reference you provide is nice and informative, but there have been much earlier papers advocating this, that one is very late to the party. There are:
- https://arxiv.org/abs/1805.11897 arguably the first to mention it
- https://arxiv.org/abs/2002.03229 (whose `log` preconditioning we use)
- a bit later, https://proceedings.neurips.cc/paper/2020/hash/ec24a54d62ce57ba93a531b460fa8d18-Abstract.html

In particular, OTT was already implementing implicit differentiation w.r.t. any parameter of the problem in April 2021:
https://github.com/google-research/ott/commit/3f57495f0de5c6ece7d20cbfacb034d9b72cbd94

Thanks again for your comment!

**Comment by marcocuturi** [comment]
Thanks for the kind word  @jaschau!!!
